src/Numerics/LinearSolvers/BatchedGeneralizedMinimalResidualSolver.jl

```diff
@@ -0,0 +1,824 @@
+    initialize_QR!(gmres, I)
+    update_QR!(1, gmres, I)
+    solve_optimization!(1, gmres, I)
+end
+
+"""
+gmres_update_kernel!(i, gmres, I)
+
+# Description
+kernel that calls
+- update_arnoldi!
+- update_QR!
+- solve_optimization!
+Which is the heart of the gmres algorithm
+
+# Arguments
+- `i`: (int) interation index
+- `gmres`: (struct) gmres struct
+- `I`: (int) thread index
+
+# Return
+kernel object from KernelAbstractions
+"""
+@kernel function gmres_update_kernel!(i, gmres)
+    I = @index(Global)
+    update_arnoldi!(i, gmres, I)
+    update_QR!(i, gmres, I)
+    solve_optimization!(i, gmres, I)
+end
+
+"""
+construct_solution_kernel!(i, gmres)
+
+# Description
+given step i of the gmres iteration, constructs the "best" solution of the linear system for the given Krylov subspace
+
+# Arguments
+- `i`: (int) gmres iteration
+- `gmres`: (struct) gmres struct
+
+# Return
+kernel object from KernelAbstractions
+"""
+@kernel function construct_solution_kernel!(i, gmres)
+    M, I = @index(Global, NTuple)
+    tmp = zero(eltype(gmres.b))
+    @inbounds for j in 1:i
+        tmp += gmres.Q[M, j, I] * gmres.sol[j, I]
+    end
+    gmres.x[M, I] += tmp # since previously gmres.x held the initial value
+end
+
+# Configuration for Kernels
+"""
+initialize_gmres!(gmres; ndrange = gmres.n, cpu_threads = Threads.nthreads(), gpu_threads = 256)
+
+# Description
+Uses the initialize_gmres_kernel! for initalizing
+
+# Arguments
+- `gmres`: (struct) [OVERWRITTEN]
+
+# Keyword Arguments
+- `ndrange`: (int) or (tuple) thread structure to iterate over
+- `cpu_threads`: (int) number of cpu threads. default = Threads.nthreads()
+- `gpu_threads`: (int) number of gpu threads. default = 256
+
+# Return
+event. A KernelAbstractions object
+"""
+function initialize_gmres!(
+    gmres::BatchedGeneralizedMinimalResidual;
+    ndrange = gmres.n,
+    cpu_threads = Threads.nthreads(),
+    gpu_threads = 256,
+)
+    if isa(gmres.b, Array)
+        kernel! = initialize_gmres_kernel!(CPU(), cpu_threads)
+    else
+        kernel! = initialize_gmres_kernel!(CUDA(), gpu_threads)
+    end
+    event = kernel!(gmres, ndrange = ndrange)
+    return event
+end
+
+"""
+gmres_update!(i, gmres; ndrange = gmres.n, cpu_threads = Threads.nthreads(), gpu_threads = 256)
+
+# Description
+Calls the gmres_update_kernel!
+
+# Arguments
+- `i`: (int) iteration number
+- `gmres`: (struct) gmres struct
+
+# Keyword Arguments
+- `ndrange`: (int) or (tuple) thread structure to iterate over
+- `cpu_threads`: (int) number of cpu threads. default = Threads.nthreads()
+- `gpu_threads`: (int) number of gpu threads. default = 256
+
+# Return
+event. A KernelAbstractions object
+"""
+function gmres_update!(
+    i,
+    gmres;
+    ndrange = gmres.n,
+    cpu_threads = Threads.nthreads(),
+    gpu_threads = 256,
+)
+    if isa(gmres.b, Array)
+        kernel! = gmres_update_kernel!(CPU(), cpu_threads)
+    else
+        kernel! = gmres_update_kernel!(CUDA(), gpu_threads)
+    end
```

Comment:
Did you see a performance improvement by changing the number of threads here?  @vchuravy should we be changing the number of threads based on the backend like this?  I am not sure we do that in other parts of the code.